python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -306,6 +200,192 @@ def sensor_fn(context):
+    source_asset_defs_by_key = (
+        repository_def.source_assets_by_key  # pylint: disable=protected-access
+    )
+    upstream: Mapping[AssetKey, Set[AssetKey]] = _get_upstream_mapping(
+        selection=asset_selection,
+        assets=asset_defs_by_key.values(),
+        source_assets=source_asset_defs_by_key.values(),
+    )
+
+    should_materialize: Set[AssetKey] = set()
+    newly_consumed_storage_ids_by_asset_key_str: Dict[str, int] = {}
+
+    # sort the assets topologically so that we process them in order
+    toposort_assets = list(toposort.toposort(upstream))
+    # unpack the list of sets into a list and only keep the ones we are monitoring
+    toposort_assets = [
+        asset for layer in toposort_assets for asset in layer if asset in upstream.keys()
+    ]
+
+    instance_queryer = CachingInstanceQueryer(instance)
+
+    # determine which assets should materialize based on the materialization status of their
+    # parents
+    for current_asset_key in toposort_assets:
+        current_asset_cursor = latest_consumed_storage_ids_by_asset_key_str.get(
+            str(current_asset_key)
+        )
+
+        parent_update_records = _get_parent_updates(
+            current_asset=current_asset_key,
+            parent_assets=upstream[current_asset_key],
+            cursor=current_asset_cursor,
+            will_materialize_set=should_materialize,
+            wait_for_in_progress_runs=wait_for_in_progress_runs,
+            instance_queryer=instance_queryer,
+        )
+
+        condition = all if wait_for_all_upstream else any
+        if condition(
+            materialization_status for materialization_status, _ in parent_update_records.values()
+        ):
+            should_materialize.add(current_asset_key)
+
+            # get the cursor value by selecting the max of all the candidates. If we're using a
+            # sharded event log storage, compare timestamps, otherwise compare storage ids. See
+            # cursor_compare_idx for how this is determined
+            cursor_update_candidates = [
+                cursor_val for _, cursor_val in parent_update_records.values() if cursor_val
+            ] + ([current_asset_cursor] if current_asset_cursor else [])
+            if cursor_update_candidates:
+                newly_consumed_storage_ids_by_asset_key_str[str(current_asset_key)] = max(
+                    cursor_update_candidates
+                )
+
+    run_requests = []
+    if len(should_materialize) > 0:
+        run_requests = [
+            RunRequest(run_key=None, asset_selection=list(should_materialize), tags=run_tags)
+        ]
+    else:
+        run_requests = []
+
+    return run_requests, newly_consumed_storage_ids_by_asset_key_str
+
+
+def _serialize_cursor_dict(cursor_dict: Dict[str, int]) -> str:
+    return json.dumps(cursor_dict)
+
+
+def _deserialize_cursor_dict(cursor: str) -> Dict[str, int]:
+    return json.loads(cursor)
+
+
+class CachingInstanceQueryer:
```

Comment:
can we rename this class to emphasize the queryer is instantiated once per tick? I was a little confused by the implementation until I realized this

Maybe `TickInstanceQueryer`?

Reply:
👍